src/core/project/qgsprojectelevationproperties.cpp

```diff
@@ -108,8 +108,13 @@ QgsAbstractTerrainProvider *QgsProjectElevationProperties::terrainProvider()
 
 void QgsProjectElevationProperties::setTerrainProvider( QgsAbstractTerrainProvider *provider )
 {
-  if ( mTerrainProvider.get() == provider )
+  if ( mTerrainProvider.get()->equals( provider ) )
```

Comment:
I'd propose a slightly different approach here -- if someone calls setTerrainProvider with a particular provider instance, then that instance should become the terrain provider, even if it's identical to the current one. 

There's also a risk of a crash here if mTerrainProvider is nullptr or provider is nullptr.

So I'd go with something like:

```
  if ( mTerrainProvider.get() == provider )
    return;

  const bool hasChanged = ( provider && mTerrainProvider ) ? !mTerrainProvider->equals( provider ) : ( static_cast< bool >( provider ) != static_cast< bool >( mTerrainProvider.get() );

   mTerrainProvider.reset( provider );
   if ( hasChanged )
     emit changed();
```

Reply:
Thanks. Changed.